### PR TITLE
Config ui

### PR DESCRIPTION
**Pull request by CTaPByK**
I hope this pull request is better.

### CODE REVIEW

File: entity_browser.routing.yml [hunk 1]
```diff
@@ -1,2 +1,28 @@
...
+  requirements:
+    _permission: 'administer entity browsers'
+
+entity.entity_browser.collection:
+  path: '/admin/config/content/entity_browser/list'
+  defaults:
+    _entity_list: entity_browser
+    _title: 'Entity Browser List'
+  requirements:
+    _permission: 'aminister entity browser'
```
Comment (slashrsm) [inline entity_browser.routing.yml]: Permission name is "administer entity browsers".

File: js/entity_browser.widgets.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  Drupal.behaviors.entityBrowserWidgets = {
+    attach: function () {
+      $('#edit-widget').on('change', function(e) {
+        e.preventDefault();
+        location.reload(true);
+      });
+    }
+  };
+
+}(jQuery, Drupal, drupalSettings));
```
Comment (slashrsm) [inline js/entity_browser.widgets.js]: Missing newline.

File: src/Controllers/EntityBrowserListBuilder.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityListBuilder;
+
+
+/**
+ * Provides a list controller for entity browser.
+ *
+ * @ingroup entity_browser
+ */
+class EntityBrowserListBuilder extends EntityListBuilder {
```
Comment (slashrsm) [inline src/Controllers/EntityBrowserListBuilder.php:19]: List builder needs to be added to EntityBrowser annotation. 

I get this error when trying to navigate to /admin/config/content/entity_browser/list:

```
Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException: The "entity_browser" entity type did not specify a list_builder handler. in Drupal\Core\Entity\EntityTypeManager->getHandler() (line 238 of /var/www/core/lib/Drupal/Core/Entity/EntityTypeManager.php).
```

File: src/Form/GeneralInfoConfig.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    /** @var \Drupal\entity_browser\EntityBrowserInterface $entity_browser */
+    $entity_browser = $form_state->getTemporaryValue('wizard')['entity_browser'];
+    $entity_browser->setName($form_state->getValue('id'))
+      ->setLabel($form_state->getValue('label'))
+      ->setDisplay($form_state->getValue('display'))
+      ->setWidgetSelector($form_state->getValue('widget_selector'))
+      ->setSelectionDisplay($form_state->getValue('selection_display'));
+
+    // TODO must be able to select single widget multiple times. We are adding
```
Comment (slashrsm) [inline src/Form/GeneralInfoConfig.php]: This part can be removed from here. We are adding widgets in a separate step.

File: src/Form/WidgetsConfig.php [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+   * @var \Drupal\entity_browser\WidgetManager
+   */
+  protected $widgetManager;
+
+  /**
+   * Tempstore Factory for keeping track of values in each step of the wizard.
+   *
+   * @var \Drupal\user\SharedTempStoreFactory
+   */
+  protected $tempStore;
```
Comment (slashrsm) [inline src/Form/WidgetsConfig.php:30]: This seems to be unused variable. Can be removed.

**Comment by slashrsm** [comment]
Some functions in pull request are missing PHPDoc blocks. Let's add them. 

I was testing widgets page a bit and I noticed few problems:
- If I was editing an existing entity browser and changed configuration of a widget those changes weren't saved. 
- Selecting widget plugin form drop-down won't update the table. I had to go to previous step and back to see it.
- All widgets get weight set to 0. Weights should respect positions in the draggable table.

File: js/entity_browser.widgets.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+(function ($, Drupal, drupalSettings) {
+
+  "use strict";
+
+  /**
+   * Registers behaviours related to config widgets page.
+   */
+  Drupal.behaviors.entityBrowserWidgets = {
+    attach: function () {
+      $('#edit-widget').on('change', function(e) {
```
Comment (slashrsm) [inline js/entity_browser.widgets.js]: This is not needed and is actually messing with AJAX if it is done correctly. One of the steps that I needed to do was to remove this entirely.

File: src/Form/WidgetsConfig.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+      /** @var \Drupal\entity_browser\WidgetInterface $widget */
+      $default_widgets[] = $widget->id();
+    }
+    $form['widget'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Add widget plugin'),
+      '#options' => ['_none_' => '- ' . $this->t('Select a widget to add it') . ' -'] + $widgets,
+      '#ajax' => [
+        'callback' => [get_class($this), 'addWidgetCallback'],
+        'wrapper' => 'widgets',
```
Comment (slashrsm) [inline src/Form/WidgetsConfig.php:81]: It turns out this ID doesn't exist at all. When AJAX framework tries to insert rebuilt table it doesn't know where to put it. 

We either need to make this ID available or change it to something that exists. I propose first option. See further explanation below.

File: src/Form/WidgetsConfig.php [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        'event' => 'change'
+      ],
+      '#executes_submit_callback' => TRUE,
+      '#submit' => [[get_class($this), 'submitAddWidget']],
+      '#limit_validation_errors' => [['widget']],
+    ];
+    $form_state->unsetValue('widget');
+
+    $form['widgets'] = [
```
Comment (slashrsm) [inline src/Form/WidgetsConfig.php:90]: I propose that we add another wrapper around the table. We can do this by adding container around it. Put this above this call:

```
    $form['widgets'] = [
      '#type' => 'container',
      '#attributes' => ['id' => 'widgets'],
    ];
```

and then change this assignment to:

```
$form['widgets']['table'] = [
  '#type' => 'table',
...
```

You will also need to fix pieces that modify table below in the same function.

Drupal render arrays magic. :)

File: src/Entity/EntityBrowser.php [hunk 9]
```diff
@@ -26,8 +26,18 @@
...
+ *     },
+ *     "access" = "Drupal\entity_browser\EntityBrowserAccessControlHandler",
+ *     "list_builder" = "Drupal\entity_browser\Controllers\EntityBrowserListBuilder",
+ *     "wizard" = {
+ *       "add" = "Drupal\entity_browser\Wizard\EntityBrowserWizardAdd",
+ *       "edit" = "Drupal\entity_browser\Wizard\EntityBrowserWizard"
  *     }
  *   },
+ *   links = {
+ *     "canonical" = "/admin/config/content/entity_browser/{entity_browser}",
```
Comment (slashrsm) [inline src/Entity/EntityBrowser.php:40]: "edit-form" needs to be added too as list builder depends on it. I suspect the same with "delete-form" (which we'll need to add in this case).

File: src/Form/WidgetsConfig.php [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    /** @var \Drupal\entity_browser\EntityBrowserInterface $entity_browser */
+    $entity_browser = $form_state->getTemporaryValue('wizard')['entity_browser'];
+    /** @var \Drupal\entity_browser\WidgetInterface $widget */
+    foreach ($entity_browser->getWidgets() as $uuid => $widget) {
+      $widget->submitConfigurationForm($form, $form_state);
```
Comment (slashrsm) [inline src/Form/WidgetsConfig.php:221]: Weights are now saved correctly, but we're still missing other configuration (label, widget's own config - upload location on upload widget for example).

**Comment by killua99** [comment]
This pull support the Display View to handle values like: width: '80%' ? For modal display this is very useful

**Comment by slashrsm** [comment]
This was committed. See https://www.drupal.org/node/2512702#comment-10872848 for the patch that went in. Thank you!

**Pull request closed without merging**